### PR TITLE
[configure]: support for profiles

### PR DESCRIPTION
**Pull request by gares**
Profiles are a way to specify in one go a set of configure flags.
Profiles are a way to have all developers use a sane set of flags.

This PR is made of 2 commits, the first one refactors the code
while the latter adds a profile. I picked my options of choice, but I'm open to suggestions.
```
./configure -help
…
  -profile <profile>          Sets a bunch of flags. Supported profiles: 
     devel = -local -annot -bin-annot -warn-error
…
```

The code lets one easily add profiles later on, so they truly belong to different PRs.
For example I can imagine a `bisect` profile that never builds the coqide even if gtk is detected.

On a side note, I think .merlin should be generated (maybe just partially) by `./configure` so that the selected flags go in there too, but this is for another PR.

### CODE REVIEW

**Review by ejgallego — COMMENTED** [review]
I'm OK with this. I am not sure thou if we would like to move out of profiles flags such as `configdir` `mandir` or even some other flags should IMO be removed ` lablgtkdir` `camlp5dir` etc... in favor of a proper ocamlfind setup.

File: configure.ml [hunk 1]
```diff
@@ -246,131 +341,107 @@ let get_ide = function
...
-  "-force-caml-version", Arg.Set Prefs.force_caml_version,
+  "-force-caml-version", arg_set (fun p force_caml_version -> { p with force_caml_version }),
     " Force OCaml version";
-  "-force-findlib-version", Arg.Set Prefs.force_findlib_version,
+  "-force-findlib-version", arg_set (fun p force_findlib_version -> { p with force_findlib_version }),
     " Force findlib version";
-  "-warn-error", Arg.Set Prefs.warn_error,
+  "-warn-error", arg_set (fun p warn_error -> { p with warn_error }),
     " Make OCaml warnings into errors";
   "-camldir", Arg.String (fun _ -> ()),
```
Comment (gares) [inline configure.ml]: This is fishy, but it is not my fault ;-)

**Comment by gares** [comment]
@ejgallego I agree, not everything makes sense to be in a profile, but some of the things you mention may. E.g. `-profile windows-installer`

**Review by SkySkimmer — COMMENTED** [review]
If I understand correctly this supports doing `./configure -profile devel -native-compiler no` (for instance) but the order is significant such that `./configure -native-compiler no -profile devel` resets native-compiler to the default. So maybe we should error when -profile appears but is not first (eg check `!prefs == default` in `arg_profile`).

Alternatively we could try to be more generic but I'm not sure that's worth the bother.

**Comment by gares** [comment]
> the order is significant

yes, things are imperative. Not sure we want to be strict, it is already clear that some options are not really part of profiles. When we remove them, the check will be silly.

So I'd stay loose.

**Comment by gares** [comment]
My main motivation, that I'll illustrate at the next WG, is that when an option is added (that is relevant to developers) one also adds it to the profile.

E.g. I missed the `-warn-error` and I complained (to @maximedenes) what is this detected by travis and not my PC! ... well, with profiles I hope it won't happen again.

**Comment by gares** [comment]
All flags should have a `off` / `on` so that one can override a profile.

**Comment by gares** [comment]
@SkySkimmer do you mind reviewing this PR?

File: configure.ml [hunk 2]
```diff
@@ -235,6 +235,101 @@ let _ = if not (dir_exists "bin") then Unix.mkdir "bin" 0o755
...
+}
+let devel_doc = "-local -annot -bin-annot -warn-error yes"
+
+let get = function
+  | "devel" -> devel
+  | s -> raise (Arg.Bad ("profile name expected instead of "^s))
+
+let doc =
+  "<profile> Sets a bunch of flags. Supported profiles:
+     devel = " ^ devel_doc
```
Comment (SkySkimmer) [inline configure.ml]: A more generic way to do this would be to have an associative list of `profile key * (profile value * profile doc)`.

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
# Requested change
I can't agree with having -profile silently override any previous options even when they're unrelated to the profile. At least make it clear in the usage that it's not "sets the following flags: -local ..." but rather "resets everything to some well-chosen values including -local ...".

# Other comments
I don't see the point of the big record but I wouldn't block over it.

What I would have done would look more like (without the change to a record):
```ocaml
let devel () =
  local := true;
  bin_annot := true;
  ...
```
which seems no more complicated than `devel = { default with ... }` and doesn't override previous options unless they're part of the profile.

**Comment by gares** [comment]
Done: now `-profile x` simply overrides its set of flags, not the entire state, hence `.... -profile x ...` is semantically equivalent to `.... -whatver -flag -in -the -profile ...`

**Comment by gares** [comment]
> I don't see the point of the big record

I wanted all options to be in in one place, so let one manipulate sets of them.
The "options are global refs" model is not very declarative IMO.

Finally, I think we should reify the entire thing, so to keep in sync the doc and the semantics of all flags.
While this is for another PR, I think it is simpler if all options are in one data structure.

File: configure.ml (hunk 2 above)
Comment (gares) [reply configure.ml]: I agree, see also my comment below.

**Review by SkySkimmer — APPROVED** [review]
OK then.

**Review by maximedenes — APPROVED** [review]

**Comment by Zimmi48** [comment]
Souldn't this PR contains some update to developer documentation as well?

**Comment by Zimmi48** [comment]
Thanks!

File: configure.ml [hunk 3]
```diff
@@ -235,6 +235,101 @@ let _ = if not (dir_exists "bin") then Unix.mkdir "bin" 0o755
...
+  local = true;
+  bin_annot = true;
+  annot = true;
+  warn_error = true;
+}
+let devel_doc = "-local -annot -bin-annot -warn-error yes"
+
+let get = function
+  | "devel" -> devel
+  | s -> raise (Arg.Bad ("profile name expected instead of "^s))
```
[the remark this answers is not in the document]
Comment (gares) [reply configure.ml]: I think this makes the "usage" message being printed, and there there should be the list of supported profiles

[the remark this answers is not in the document]
Comment (maximedenes) [reply configure.ml]: Ah, I see. Good then.

**Comment by maximedenes** [comment]
I'm taking care of conflicts.

**Comment by maximedenes** [comment]
@gares what was the motivation for putting `-annot` in the `devel` profile?

**Pull request merged**